arch/risc-v/src/common/riscv_macros.S

```diff
@@ -128,3 +133,22 @@
   REGLOAD    x31, REG_X31(\out)  /* t6 */
 
 .endm
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 15
+.macro  setintstack tmp0, tmp1
+#ifndef CONFIG_ARCH_USE_SMODE
+#if CONFIG_SMP_NCPUS > 1
+  csrr    \tmp0, mhartid
```

Comment:
This code is also IMO wrong but the original code was not made by me. The problem is that SMP cpu index is not the same has the physical hart ID. And we do need both.

- CPU id is an incrementing number between 0...SMP_CPUs, needed e.g. for calculating the IRQ stack position
- hartid is the physical core number, needed for external interrupt handling (PLIC) at least

Reply:
It depends on how hardware/software design and config. In many simple design, hartid match index. If you want to support the index isn't equal to hartid. M mode also need the similar percpu struct. Since how to manage hartid is very hardware specific, that's why both riscv spec and NuttX left the detail to chip specific implementation.

Reply:
Of course, just something to keep in mind here. Can easily produce strange results.

A good example is MPFS, which has 5 harts (SiFive cores). 
Core 0 is u51 which does not have S-mode, FPU, etc. 
Cores 1-4 are u54, which have S-mode / FPU.

So running SMP on e.g. core0&1 makes handling the code difficult, because the cores have different capabilities ! This is why I run NuttX on core 4.

Reply:
it's fine to put riscv_percpu.c in the common place, because user can simply ignore it by excluding it from compiling, but it's hard to replace setintstack if it define in riscv_macros.S, that's why I suggest riscv is better to follow other arch not define setintstack in the common place.

Reply:
I agree it should be in the chip specific files, but no risc-v target implements setintstack. This basically like how it was before.

Do you think I should add this code to every target or leave it like this for now ? Because if I implement this for mpfs only all other targets that use IRQ stack (every risc-v target in fact) will break ?

Reply:
It's by design. since in the SMP case, the user must provide setintstack before they can enable interrupt stack.

Reply:
I think all (or at least many) risc-v targets use interrupt stack now even without SMP. This macro will work for them, because setting the interrupt stack was just:

exception_common:
` la sp, g_intstacktop`


This does not quite work either, because setinstack for mpfs is needed when 1 hart is used (to use the percpu structure):

exception_common:
```
#if CONFIG_SMP_NCPUS > 1
  setintstack t0, t1
#else
  la sp, g_intstacktop
#endif
```

Reply:
I think this should now be solved. Old implementations use common macro which transforms into:
la sp, g_intstacktop

For MPFS there is a specific solution